proposals/p3833.md

```diff
@@ -0,0 +1,211 @@
+
+First, this will allow us to study the resulting SemIR, and understand exactly
+how and where the efficiency explodes for expected coding patterns. We should
+use this to try to adjust the design to build in targeted common-case
+simplifications and elisions that make precise modeling of the design
+sufficiently cheap in practice.
+
+We see this pattern emerging in good ways already today in the toolchain with
+implicit conversions: when calling functions, we need to allow for implicit
+conversions. But we don't want to have an identity implicit conversion on all
+the arguments which already are the correct type. Rather than notionally
+"always" implicitly converting, we make the design itself specify that the
+implicit conversion only occurs when needed and can then have the desired
+minimal representation in SemIR.
+
+Beyond simplifying the SemIR model, it also ends up addressing an important
+aspect of the semantics -- we don't want to create unnecessary temporaries to
+initialize rather than directly initializing. The toolchain in fact follows
+similar patterns to model expression categories for similar reasons -- the
+minimal semantic model is also the desired design.
+
+We can even incorporate other kinds of design changes to allow the full fidelity
+SemIR model to be efficient, such as assuring that we can build and reuse a
+cached representation repeatedly and cheaply rather than requiring it to be
+rebuilt on each instance of a specific pattern.
+
+### Keep a full fidelity mode for any other optimizations
+
+We may still ultimately end up unable to achieve the desired SemIR efficiency
+even after making all the simplifications we can to the design itself. If and
+when we reach this point, we will still want to have the full fidelity
+implementation and be able to use it behind some option. We can use it to
+understand how the design works in complex cases and to debug unexpected
+behavior in the toolchain. This means starting with that implementation doesn't
+result in sunk cost.
+
+### Example: overloaded operators
+
+> Note: this example isn't trying to capture all the nuance of the current
+> design or suggest any change to the design. It is merely trying to provide an
+> illustration of how we might follow the proposal in practice.
+
+As an example to help illustrate this, how should the toolchain model overloaded
+operators? This proposal suggests that _initially_, the implementation should
+aim to fully model the rewrite-based dispatch through interfaces in the prelude.
+That is, each use of `x + y` should turn into roughly equivalent SemIR as would
+be used to model the rewritten semantics of `x.(Core.AddWith(typeof(y)).Op)(y)`.
+This in turn would dispatch to an exact-type implementation which would provide
+any implicit conversions, and so-on.
+
+To be precise, the expectation is that the SemIR for `x + y` should as a
+consequence model all of:
+
+-   Looking up the `Core.AddWith` interface.
```

Comment:
Maybe:
```suggestion
-   Looking up the `AddWith` interface within the standard library package.
```
In particular, an unqualified lookup of the name `Core` isn't part of the design. (Performing a qualified lookup for the interface names wasn't my intent in #1083 either, but I don't think that's an important distinction, and I'd be OK with the rule being that you do perform that qualified lookup. But I'd be pretty strongly opposed to using an unqualified lookup for the name `Core`.)

Reply:
Ah, now I understand the confusion when we've discussed this in the past.

Definitely not suggesting we should be doing arbitrary unqualified lookup. I agree that would be bad.

Really, this seems to suggest that there is a fundamental gap -- we should have some unambiguous way of looking up a name in the `Core` package without doing that name lookup. We have `package.Name`, we kind of need `(package Core).Name` but with some much better syntax. Is there a way I could frame this in pseudo syntax here?

Reply:
I'd be fine with the syntax you suggested, eg:
```suggestion
-   Looking up the `(package Core).AddWith` interface, where `package Core` is pseudo-syntax that directly names the `Core` package.
```

Reply:
I've added a slightly different imagined syntax and clarified the behavior and that these are all imagined syntaxes.